apps/src/music/blockly/MusicBlocklyWorkspace.ts

```diff
@@ -454,23 +455,32 @@ export default class MusicBlocklyWorkspace {
     const codeCopy = JSON.parse(JSON.stringify(code));
 
     Blockly.serialization.workspaces.load(codeCopy, this.workspace);
-
-    if (
-      this.toolbox?.addFunctionCalls &&
-      this.toolbox?.type === 'flyout' &&
-      this.blockMode === BlockMode.SIMPLE2
-    ) {
-      this.generateFunctionCallBlocks();
-    }
   }
 
   // For each function body in the current workspace, add a function call
-  // block to the toolbox.
-  generateFunctionCallBlocks() {
+  // block to the toolbox. Also add a function defintion block, if required.
+  generateFunctionBlocks() {
+    const blockList: ToolboxItemInfo[] = [];
+
+    if (this.toolbox?.addFunctionDefinition) {
+      blockList.push({
+        kind: 'block',
+        type: BLOCK_TYPES.procedureDefinition,
+        fields: {
+          NAME: musicI18n.blockly_functionNamePlaceholder(),
+        },
+      });
+    }
+
     const allFunctions: GoogleBlockly.serialization.procedures.State[] = [];
 
     (this.workspace?.getTopBlocks() as ProcedureBlock[])
-      .filter(block => block.type === 'procedures_defnoreturn')
+      .filter(
+        // Disregarding unrendered blocks prevents duplicating call blocks
+        // for a defintion block that was just pulled from the toolbox.
```

Comment:
The double negative in "Disregarding unrendered" could be avoided with "Only including rendered blocks" or similar. :) 

What exactly does "just pulled from the toolbox" mean?  Is it a block that is still being dragged but not yet placed in the workspace?

Also small typo in "definition".

Reply:
Great question! It prompted me to dig a little deeper. 

What actually happens is that when you drag a block out, an insertion marker is also created (which provides a visual indicator of where a block will land when you stop dragging it). This "block" isn't rendered until you're within the snap radius of another block. We don't want to count the insertion marker because it results in additional call block being created for the same procedure model.

I had picked up on there being an un-rendered block in the mix but I hadn't figured out quite why. I've updated the comment to clarify and switched to checking `block.isInsertionMarker()` which is what we actually care about.